lesson 4/Tensor/Tensor/tensor.cpp

```diff
@@ -0,0 +1,239 @@
+	for (std::size_t i = 0; i < size; ++i) {
+		data[i] = copyObj.data[i];
+	}
+}
+
+Tensor1D::Tensor1D(Tensor1D&& moveObj): Tensor1D() {
+	swap(*this, moveObj);
+}
+
+Tensor1D& Tensor1D::operator=(const Tensor1D& copyAss) {
+	Tensor1D tmp(copyAss);
+	swap(*this, tmp);
+	return *this;
+}
+
+Tensor1D& Tensor1D::operator=(Tensor1D&& moveAss) {
+	Tensor1D tmp(std::move(moveAss));
+	swap(*this, moveAss);
+	return *this;
+}
+void swap(Tensor1D& left, Tensor1D& right) {
+	std::swap(left.data, right.data);
+	std::swap(left.size, right.size);
+	std::swap(left.shape, right.shape);
+}
+
+void Tensor1D::transpose() {
+	Tensor1D tmp(*this);
+	for (std::size_t k = 0, i = size - 1; k < size; ++k, --i) {
+			data[k] = tmp.data[i];
+	}
+}
+
+std::array<std::size_t, 2> Tensor1D::getShape() {
+	return shape;
+}
+
+std::vector<double> Tensor1D::flatten() const {
+	std::vector<double> flattend(size);
+	std::copy(&data[0], &data[size], flattend.begin());
+	return flattend;
+} 
+
+void Tensor1D::printElements() const{
+	for (std::size_t i = 0; i < size; ++i) {
+		std::cout << data[i] << " " << std::flush;
+	}
+	std::cout << std::endl;
+}
+
+/// /////////////// TENSOR 2 D /////////////////////////////////////
+
+void Tensor2D::setValue(const double val) {
+	for (std::size_t i = 0; i < shape[0]; ++i) {
+		for (std::size_t k = 0; k < shape[1]; ++k) {
+			data[i][k] = val;
+		}
+	}
+}
+
+void Tensor2D::allocate2D(const std::size_t rowSize, const std::size_t columnSize) {
+	data = new double* [rowSize];
+	for (std::size_t i = 0; i < rowSize; ++i) {
+		data[i] = new double[columnSize];
+	}
+}
+
+void Tensor2D::allocate2D(const std::size_t rowSize, const std::size_t columnSize, const double defaultVal) {
+	data = new double* [rowSize];
+	std::vector<double> val(columnSize, defaultVal);
+	for (std::size_t i = 0; i < rowSize; ++i) {
+		data[i] = new double[columnSize];
+		std::copy(val.begin(), val.end(), &data[i][0]);
+	}
+}
+
+Tensor2D::Tensor2D(std::initializer_list<std::initializer_list<double>> data_) {
+
+	auto* it = data_.begin();
+	std::size_t in_size = it->size(); // {{in_size}, {in_size}}
+	
+	while (++it != data_.end()) {
+		if (in_size != it->size()) {
+			throw WrongShapeTensorException();
+		}
+	}
+	
+	shape = { data_.size(), in_size };
+	allocate2D(shape[0], shape[1]);
+
+	std::size_t i = 0;
+	for (const auto& tmp : data_) {
+		std::copy(tmp.begin(), tmp.end(), data[i++]);
+	}
+}
+
+Tensor2D::Tensor2D(const std::size_t m, const std::size_t n) : shape{ m, n } {
+	allocate2D(m, n);
+}
+
+Tensor2D::Tensor2D(const std::size_t m, const std::size_t n, const double defaultVal): shape{ m, n } {
+	allocate2D(m, n, defaultVal);
+}
+Tensor2D::~Tensor2D() {
+	for (std::size_t i = 0; i < shape[0]; ++i) {
+		delete[] data[i];
+	}
+	delete[] data;
+}
+
+Tensor2D::Tensor2D(const Tensor2D& copyObj) {
+	shape = copyObj.shape;
+	allocate2D(shape[0], shape[1]);
+
+	for (std::size_t i = 0; i < shape[0]; ++i) {
+		std::copy(&copyObj.data[i], &copyObj.data[shape[0]], &data[i]);
```

Comment:
Your indexing is wrong: You need to take the address of the first element of the i-th row of `copyObj.data` and the address of the last element of the same line. This issue affects on Tensor3D construction too